File: my-folder/0820-find-eventual-safe-states/solution.cpp
class Solution {
public:
    
    bool dfs(int node,vector<vector<int>>& graph, vector<int>& state)
    {
        if (state[node] != 0) {
            return state[node] == 2;
        }

        state[node] = 1;

        for (int nei : graph[node]) {
            if (state[nei] == 1 || !dfs(nei, graph, state)) {
                return false;
            }
        }

        state[node] = 2;
        return true;
    }
    
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        
        vector<int> output;
        vector<int> states(n,0); 
        for (int i=0; i<n;i++)
        {
            if ( dfs(i,graph,states) )
                output.push_back(i);
        }

        return output;
    }
};

File: my-folder/0820-find-eventual-safe-states/solution.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        # 0--unsafe, 1--safe, 2--terminal, -1-->unvisited, -2-->visiting
        states = defaultdict(int)
        for i in range(n): states[i] = -1

        def dfs(node):
            if states[node]!=-1:return 
            states[node] = -2
            if len(graph[node])==0:
                states[node] = 2
            for child in graph[node]:
                dfs(child)
            is_safe = True
            for child in graph[node]:
                if states[child]!=1:
                    states[node] = 0
                    is_safe = False
                    break
            if is_safe:
                states[node] = 1
        for i in range(n):
            dfs(i)
        
        ans = []
        for i in range(n):
            if states[i]==1:
                ans.append(i)
        return sorted(ans)
            
